lib/protocol/protocol.go

```diff
@@ -372,39 +368,35 @@ func (c *rawConnection) readMessage() (message, error) {
 func (c *rawConnection) readMessageAfterHeader(hdr Header) (message, error) {
 	// First comes a 4 byte message length
 
-	if len(c.readerBuf) < 4 {
-		c.readerBuf = make([]byte, 4)
-	}
-	if _, err := io.ReadFull(c.cr, c.readerBuf[:4]); err != nil {
+	buf := buffers.get(4)
+	if _, err := io.ReadFull(c.cr, buf); err != nil {
 		return nil, fmt.Errorf("reading message length: %v", err)
 	}
-	msgLen := int32(binary.BigEndian.Uint32(c.readerBuf))
+	msgLen := int32(binary.BigEndian.Uint32(buf))
 	if msgLen < 0 {
 		return nil, fmt.Errorf("negative message length %d", msgLen)
 	}
 
 	// Then comes the message
 
-	if len(c.readerBuf) < int(msgLen) {
-		c.readerBuf = make([]byte, msgLen)
-	}
-	if _, err := io.ReadFull(c.cr, c.readerBuf[:msgLen]); err != nil {
+	buf = buffers.upgrade(buf, int(msgLen))
```

Comment:
Why not put and get? I can't see the code for the pool implementation over the phone atleast.